Firestore/core/src/firebase/firestore/local/leveldb_migrations.cc

```diff
@@ -241,6 +246,65 @@ void EnsureSentinelRows(leveldb::DB* db) {
   transaction.Commit();
 }
 
+// Helper to add an index entry iff we haven't already written it (as determined
+// by the provided cache).
+void EnsureCollectionParentRow(LevelDbTransaction* transaction,
+                               MemoryCollectionParentIndex* cache,
+                               const ResourcePath& path) {
+  if (cache->Add(path)) {
+    std::string collection_id = path.last_segment();
+    ResourcePath parent_path = path.PopLast();
+
+    std::string key =
+        LevelDbCollectionParentKey::Key(collection_id, parent_path);
+    std::string empty_buffer;
+    transaction->Put(key, empty_buffer);
+  }
+}
+
+/**
+ * Migration 6.
+ *
+ * Creates appropriate LevelDbCollectionParentKey rows for all collections
+ * of documents in the remote document cache and mutation queue.
+ */
+void EnsureCollectionParentIndex(leveldb::DB* db) {
+  LevelDbTransaction transaction(db, "Ensure Collection Parents Index");
+
+  MemoryCollectionParentIndex cache;
+
+  // Index existing remote documents.
+  std::string documents_prefix = LevelDbRemoteDocumentKey::KeyPrefix();
+  auto it = transaction.NewIterator();
+  it->Seek(documents_prefix);
+  LevelDbRemoteDocumentKey documents_key;
+  for (; it->Valid() && absl::StartsWith(it->key(), documents_prefix);
+       it->Next()) {
+    HARD_ASSERT(documents_key.Decode(it->key()),
+                "Failed to decode document key");
+
+    const ResourcePath& path = documents_key.document_key().path();
+    EnsureCollectionParentRow(&transaction, &cache, path.PopLast());
+  }
+
+  // Index existing mutations.
+  std::string mutations_prefix = LevelDbDocumentMutationKey::KeyPrefix();
+  it = transaction.NewIterator();
+  it->Seek(mutations_prefix);
+  LevelDbDocumentMutationKey key;
+  for (; it->Valid() && absl::StartsWith(it->key(), mutations_prefix);
+       it->Next()) {
+    HARD_ASSERT(key.Decode(it->key()),
+                "Failed to decode document-mutation key");
+
+    const ResourcePath& path = key.document_key().path();
+    EnsureCollectionParentRow(&transaction, &cache, path.PopLast());
```

Comment:
Fair point.  Done!  Note: I left this as-is on the other platforms since the remote document and mutation paths come out of persistence as ResourcePath, so we'd have to go out of our way to wrap the path in a DocumentKey and then unwrap it again.